src/main/java/com/google/step/snippet/servlets/SearchServlet.java

```diff
@@ -69,50 +77,15 @@ public void doGet(HttpServletRequest request, HttpServletResponse response)
       response.getWriter().println("Invalid Query");
       return;
     }
-
     String query = encodeValue(param);
     List<Card> allCards = new ArrayList<>();
-
-    // TODO: after implementing scraping, consider changing getLink calls to a
-    // for-loop
-
-    /*
-     * Send request to retrieve card content through w3Schools site link from Google
-     * CSE
-     */
-    String w3Link = getLink(W3_CSE_ID, query);
-    if (w3Link != null) {
-      W3SchoolsClient client = new W3SchoolsClient();
-      Card w3Card = client.search(w3Link);
-      if (w3Card != null) {
-        allCards.add(w3Card);
-      }
-    }
-
-    /*
-     * Send request to retrieve card content through StackOverflow site link from
-     * Google CSE
-     */
-    String stackLink = getLink(STACK_CSE_ID, query);
-    if (stackLink != null) {
-      StackOverflowClient stackClient = new StackOverflowClient();
-      Card stackCard = stackClient.search(stackLink);
-      if (stackCard != null) {
-        allCards.add(stackCard);
-      }
-    }
-
-    /*
-     * Send request to retrieve card content through GeeksForGeeks site link from
-     * Google CSE
-     */
-    String geeksLink = getLink(GEEKS_CSE_ID, query);
-    if (geeksLink != null) {
-      GeeksForGeeksClient geeksClient = new GeeksForGeeksClient();
-      Card geeksCard = geeksClient.search(geeksLink);
-      if (geeksCard != null) {
-        allCards.add(geeksCard);
+    for (Client client : clients) {
+      String link = getLink(client.getCseId(), query);
+      Card card = client.search(link);
+      if (card == null) {
+        break;
```

Comment:
I don't think you want to break - the first client that fails prevents any other card from being added. Perhaps just add the card if it's not null?